File: advent-of-code-2021/src/advent_of_code_2021/10/part2.clj
(ns advent-of-code-2021.10.part2
  (:require [advent-of-code-2021.10.part1 :as part1 :refer [example input]]
            [clojure.string :as str]))

(defn find-incompletes
  "Given a line, remove matching brackets and return list of incomplete openers"
  [line]
  (let [c (part1/first-closer line)
        idx-c (str/index-of line (str c))
        o (part1/match-bracket c)
        idx-o (dec idx-c)]
    ;; We've matched all the closing brackets, return list of extra openers
    (if (nil? c) line
        ;; Otherwise, remove current matching pair from the line and run again
        (find-incompletes (part1/str-splice line idx-o (inc idx-c))))))

#_(find-incompletes (first example))

(defn complete-line
  "Given a string of opening brackets, return their closing string"
  [incompletes]
  (->> incompletes
      (map part1/match-bracket)
      (reverse)
      (apply str)))

#_(complete-line "(([{{<")

(defn score-line
  "Given a string of closing brackets, return their point value"
  [closers]
  (reduce (fn [acc c]
            (+ (* acc 5)
               (case c \) 1 \] 2 \} 3 \> 4)))
          0 closers))

#_(score-line "])}>")

(defn run [lines]
  (let [results (->> lines
                     (filter #(nil? (part1/find-corruption %)))
                     (map find-incompletes)
                     (map complete-line)
                     (map score-line)
                     (sort))
        num (count results)
        mid (Math/floor (/ num 2))]
    (nth results mid)))

#_(println "\nDay 10 - Part 2: " (run input))

File: 2022/01/part2.js
const { readInputLines } = require("../utils");

const sample = [
  1000,
  2000,
  3000,
  null,
  4000,
  null,
  5000,
  6000,
  null,
  7000,
  8000,
  9000,
  null,
  10000
];
const input = readInputLines().map(l => l ? Number(l) : null);

const solve = () => {
  let most = [0, 0, 0];
  let current = 0;

  for (let n of [...input, null]) {
    if (n === null) {
      const index = most.findIndex(m => m < current);
      if (index > -1) {
        most[index] = current;
        most.sort((a, b) => a - b);
      }
      current = 0;
    } else {
      current += n;
    }
  }

  return most.reduce((sum, num) => sum + num);
}

console.log(solve());
